### PR TITLE
Expand `CoilImage` to allow handling of error results

### PR DESCRIPTION
**Pull request by chrisbanes**
This splits out the creation of the `ImagePainter`s to lambda parameters, back by fallback implementations.

This enables the implementation of `CoilImageWithCrossfade` to be an overload of `CoilImage` with a custom painter builder internally. Eventually this will lead to some sort of transition system, but I need more about the API.

### CODE REVIEW

File: coil/src/main/java/dev/chrisbanes/accompanist/coil/Coil.kt [hunk 1]
```diff
@@ -234,13 +72,19 @@ fun CoilImage(
     alignment: Alignment = Alignment.Center,
     contentScale: ContentScale = ContentScale.Fit,
     colorFilter: ColorFilter? = null,
+    onRequestCompleted: (RequestResult) -> Unit = emptySuccessLambda,
```
Comment (nickbutcher) [inline coil/src/main/java/dev/chrisbanes/accompanist/coil/Coil.kt:75]: Make last param for trailing lambda?

File: coil/src/main/java/dev/chrisbanes/accompanist/coil/Coil.kt [hunk 2]
```diff
@@ -264,15 +108,17 @@ fun CoilImage(
...
```
Comment (nickbutcher) [inline coil/src/main/java/dev/chrisbanes/accompanist/coil/Coil.kt:111]: ditto

File: coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+import androidx.ui.foundation.Image
+import androidx.ui.graphics.ImageAsset
+import androidx.ui.graphics.painter.Painter
+import coil.Coil
+import coil.decode.DataSource
+import coil.request.GetRequest
+import coil.request.GetRequestBuilder
+import kotlin.math.roundToInt
+
+private const val defaultTransitionDuration = 1000
```
Comment (nickbutcher) [inline coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt:43]: Pascal case

File: coil/src/androidTest/java/dev/chrisbanes/accompanist/coil/CoilTest.kt [hunk 4]
```diff
@@ -16,21 +16,214 @@
...
-        // TODO
+    fun onRequestCompleted() {
+        val results = ArrayList<RequestResult>()
+        val latch = CountDownLatch(1)
+
+        composeTestRule.setContent {
+            CoilImage(
+                request = GetRequest.Builder(ContextAmbient.current)
+                    .data(rawUri(R.raw.sample))
+                    .listener { _, _ -> latch.countDown() }
```
Comment (nickbutcher) [inline coil/src/androidTest/java/dev/chrisbanes/accompanist/coil/CoilTest.kt:69]: Can't you just omit the lambda params?

File: coil/src/androidTest/java/dev/chrisbanes/accompanist/coil/CoilTest.kt [hunk 5]
```diff
@@ -16,21 +16,214 @@
...
+                    modifier = Modifier.preferredSize(128.dp, 128.dp),
+                    onRequestCompleted = { latch.countDown() }
+                )
+            }
+        }
+
+        // Wait for the onRequestCompleted to release the latch
+        latch.await(5, TimeUnit.SECONDS)
+
+        // Assert that the whole layout is drawn red
```
Comment (nickbutcher) [inline coil/src/androidTest/java/dev/chrisbanes/accompanist/coil/CoilTest.kt:154]: stale comment

File: coil/src/androidTest/java/dev/chrisbanes/accompanist/coil/CoilTest.kt [hunk 6]
```diff
@@ -16,21 +16,214 @@
...
+    @SdkSuppress(minSdkVersion = 26) // captureToBitmap is SDK 26+
+    fun customGetPainter() {
+        val latch = CountDownLatch(1)
+
+        composeTestRule.setContent {
+            TestTag(CoilTestTags.Image) {
+                CoilImage(
+                    data = rawUri(R.raw.sample),
+                    getSuccessPainter = {
+                        // Return a custom success painter which just draws green
```
Comment (nickbutcher) [inline coil/src/androidTest/java/dev/chrisbanes/accompanist/coil/CoilTest.kt:118]: cyan

File: coil/src/androidTest/java/dev/chrisbanes/accompanist/coil/CoilTest.kt (hunk 4 above)
Comment (chrisbanes) [reply coil/src/androidTest/java/dev/chrisbanes/accompanist/coil/CoilTest.kt:69]: Unfortunately not.

File: coil/src/main/java/dev/chrisbanes/accompanist/coil/ColorMatrixImagePainter.kt [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+        drawCanvas { canvas, _ ->
+            canvas.drawImageRect(image, srcOffset, srcSize, Offset.zero, size, paint)
+        }
+
+        paintPool.release(paint)
+    }
+
+    /**
+     * Return the dimension of the underlying [Image] as it's intrinsic width and height
```
Comment (JoseAlcerreca) [inline coil/src/main/java/dev/chrisbanes/accompanist/coil/ColorMatrixImagePainter.kt:52]: its intrinsic

File: coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+ *
+ * @param data The data to load. See [GetRequestBuilder.data] for the types allowed.
+ * @param alignment Optional alignment parameter used to place the loaded [ImageAsset] in the
+ * given bounds defined by the width and height.
+ * @param contentScale Optional scale parameter used to determine the aspect ratio scaling to be
+ * used if the bounds are a different size from the intrinsic size of the loaded [ImageAsset].
+ * @param crossfadeDuration The duration of the crossfade animation in milliseconds.
+ * @param getFailurePainter Optional builder for the [Painter] to be used to draw the failure
+ * loading result. Passing in `null` will result in falling back to the default [Painter].
+ * @param modifier Modifier used to adjust the layout algorithm or draw decoration content (ex.
```
Comment (JoseAlcerreca) [inline coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt:60]: [Modifier]

File: coil/src/androidTest/java/dev/chrisbanes/accompanist/coil/CoilTest.kt [hunk 9]
```diff
@@ -16,21 +16,214 @@
...
+    @SdkSuppress(minSdkVersion = 26) // captureToBitmap is SDK 26+
+    fun customFailurePainter() {
+        val latch = CountDownLatch(1)
+
+        composeTestRule.setContent {
+            TestTag(CoilTestTags.Image) {
+                CoilImage(
+                    data = "url_which_will_never_work",
+                    getFailurePainter = {
+                        // Return a custom success painter which just draws green
```
Comment (nickbutcher) [inline coil/src/androidTest/java/dev/chrisbanes/accompanist/coil/CoilTest.kt:206]:
```suggestion
                        // Return a custom failure painter which just draws red
```

File: coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+ */
+@Composable
+fun CoilImageWithCrossfade(
+    data: Any,
+    alignment: Alignment = Alignment.Center,
+    contentScale: ContentScale = ContentScale.Fit,
+    crossfadeDuration: Int = DefaultTransitionDuration,
+    getFailurePainter: @Composable ((ErrorResult) -> Painter?)? = null,
+    loading: @Composable (() -> Unit)? = null,
+    modifier: Modifier = Modifier,
```
Comment (nickbutcher) [inline coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt:72]: First optional param

File: coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+ * @param modifier Modifier used to adjust the layout algorithm or draw decoration content (ex.
+ * background)
+ * @param onRequestCompleted Listener which will be called when the loading request has finished.
+ */
+@Composable
+fun CoilImageWithCrossfade(
+    request: GetRequest,
+    alignment: Alignment = Alignment.Center,
+    contentScale: ContentScale = ContentScale.Fit,
+    crossfadeDuration: Int = DefaultTransitionDuration,
```
Comment (nickbutcher) [inline coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt:112]: unit?

File: coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+    getFailurePainter: @Composable ((ErrorResult) -> Painter?)? = null,
+    loading: @Composable (() -> Unit)? = null,
+    modifier: Modifier = Modifier,
```
Comment (nickbutcher) [inline coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt:115]: first optional param

File: coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+            observablePainter.painter
+        } else {
+            // If the animation has finished, revert back to the default painter
+            defaultSuccessPainterGetter(result)
+        }
+    }
+}
+
+@Suppress("JoinDeclarationAndAssignment")
+private class ObservableCrossfade(
```
Comment (nickbutcher) [inline coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt:168]: This is neat!

File: coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+        Loaded, Empty
+    }
+
+    internal val Alpha = FloatPropKey()
+    internal val Brightness = FloatPropKey()
+    internal val Saturation = FloatPropKey()
+
+    fun definition(durationMs: Int) = transitionDefinition {
+        state(State.Empty) {
+            this[Alpha] = 0f
```
Comment (nickbutcher) [inline coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt:224]: I wonder if these start values should be exposed (with defaults) and/or moved to constants.

File: coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+        private set
+
+    private val animation: TransitionAnimation<CrossfadeTransition.State>
+
+    init {
+        animation = CrossfadeTransition.definition(crossfadeDuration).createAnimation(clock)
+
+        animation.onUpdate = {
+            // Animation tick, so update the painter using the current transition property values
+            painter = createCrossfadePainter(
```
Comment (JoseAlcerreca) [inline coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt:192]: creating this on every tick seems expensive. You can reuse a ImageLoadingColorMatrix in createCrossfadePainter. Can you mutate ColorMatrixImagePainter?

File: coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+): Painter {
+    // Create and update the ImageLoadingColorMatrix from the transition state
+    val matrix = ImageLoadingColorMatrix()
+    matrix.saturationFraction = saturation
+    matrix.alphaFraction = alpha
+    matrix.brightnessFraction = brightness
+    return ColorMatrixImagePainter(image, colorMatrix = matrix)
+}
+
+private inline val SuccessResult.isFromMemory
```
Comment (nickbutcher) [inline coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt:262]: A `fun` feels more appropriate here?

File: coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt (hunk 15 above)
Comment (chrisbanes) [reply coil/src/main/java/dev/chrisbanes/accompanist/coil/Crossfade.kt]: Yeah, fixed this. Had to collapse down some classes, but far less objects are created now.

**Pull request merged**